### PR TITLE
Feature/create project script

### LINKED ISSUE #47: Create script to create a project
Create a function which takes project name as input and creates our default projects activities into it using Clockify API.

### PR DESCRIPTION
**Pull request by ribeirojose**
Fix #47

### CODE REVIEW

File: services/create_project.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+
+@click.command()
+@click.option("--name", prompt="Project name", help="The name for the project.")
+# TODO: implement clients in our DB
+def create_project(name: AnyStr) -> None:
+    colors = {
+        "e": "#69b39e",
+        "t": "#09215d",
+        "n": "#FF9800",
```
Comment (lukasab) [inline services/create_project.py]: Not sure if this changes something, but this is how it is in clockify:

```suggestion
        "n": "#ff9800",
```

File: services/create_project.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        "w": "#1b60ce",
+        "c": "#8BC34A",
```
Comment (lukasab) [inline services/create_project.py]: Same here:

```suggestion
        "c": "#8bc34a",
```

File: services/create_project.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    project_name = name.upper()
+    default_activities = Project.find(0).activities
+    proj_data = {
+        "name": project_name,
+        "isPublic": "true",  # On Clockify this means the project is visible to the whole team
+        "billable": "true",
+        "color": colors[project_name.lower()[0]],
+        "tasks": [{"name": x.name} for x in default_activities.all()],
```
Comment (lukasab) [inline services/create_project.py]:
```suggestion
        "tasks": [{"name": activity.name} for activity in default_activities.all()],
```

File: services/create_project.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        json={"name": name},
+        headers=HEADERS,
+    )
+    print(f"Client {name} created.")
+    pass
+
+
+@click.command()
+@click.option("--name", prompt="Project name", help="The name for the project.")
+# TODO: implement clients in our DB
```
Comment (lukasab) [inline services/create_project.py]: Just so I know, what is this?

File: services/create_project.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        "tasks": [{"name": x.name} for x in default_activities.all()],
+    }
+    proj_response = requests.post(
+        f"{V1_API_URL}/workspaces/{WORKSPACE_ID}/projects",
+        json=proj_data,
+        headers=HEADERS,
+    )
+    proj_clockify_id = proj_response.json().get("id")
+
+    new_proj = Project.create(name=project_name, clockify_id=proj_clockify_id)
```
Comment (lukasab) [inline services/create_project.py]: Nice! I would forget to create the project in our DB.

File: services/create_project.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    new_proj.activities().sync(default_activities.map(lambda x: x.id).all())
```
Comment (lukasab) [inline services/create_project.py]: I'm not sure what this does, but:

```suggestion
    new_proj.activities().sync(default_activities.map(lambda activity: activity.id).all())
```

**Review by lukasab — CHANGES_REQUESTED** [review]
Well done. You used click :blush:!!

File: services/create_project.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        "c": "#8BC34A",
+    }
+
+    project_name = name.upper()
+    default_activities = Project.find(0).activities
+    proj_data = {
+        "name": project_name,
+        "isPublic": "true",  # On Clockify this means the project is visible to the whole team
+        "billable": "true",
+        "color": colors[project_name.lower()[0]],
```
Comment (lukasab) [inline services/create_project.py]: I think it's missing an exception for when the project doesn't start with does letters.

**Comment by lukasab** [comment]
Added this to our [wiki](http://wiki.neo.ufsc.br/index.php?title=Uso_do_Clockify).

**Comment by lukasab** [comment]
I could not run the code because it needs the _template_ project. How are you thinking about creating this project(The one with all default activities)?

**Comment by ribeirojose** [comment]
> I could not run the code because it needs the _template_ project. How are you thinking about creating this project(The one with all default activities)?

We should probably add this with the migrations, right? We should probably have a seed file to populate the db with this sort of thing. What do you think?

File: services/create_project.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    proj_data = {
+        "name": project_name,
+        "isPublic": "true",  # On Clockify this means the project is visible to the whole team
+        "billable": "true",
+        "color": colors.get(project_name[0].lower(), "#ffffff"),
+        "tasks": [{"name": activity.name} for activity in default_activities.all()],
+        "clientId": client.clockify_id,
+    }
+
+    proj_response = requests.post(
```
Comment (ribeirojose) [inline services/create_project.py]: This should be a `Project` method, so you don't have to star import everything.

**Comment by lukasab** [comment]
> > I could not run the code because it needs the _template_ project. How are you thinking about creating this project(The one with all default activities)?
> 
> We should probably add this with the migrations, right? We should probably have a seed file to populate the db with this sort of thing. What do you think @lukasab ?

I think it does make a lot of sense adding this to the project/activities migration file. What do you mean by a seed file?

File: services/create_project.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+def valid_project_name(name: str, valid_initials: List[str]) -> bool:
+    if len(name) < 4:
+        return False
+    if not (name[0].lower() in valid_initials):
+        return False
+    try:
+        proj_int = int(name[1:])
+        return True
+    except:
```
Comment (lukasab) [inline services/create_project.py]: Just to be specific about what exception you're trying to catch.

```suggestion
    except ValueError:
```

File: services/create_project.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+def create_project(name: str) -> None:
+    colors = {
+        "e": "#69b39e",
+        "t": "#09215d",
+        "n": "#ff9800",
+        "w": "#1b60ce",
+        "c": "#8bc34a",
+    }
+
+    if not valid_project_name(name, colors.keys()):
```
Comment (lukasab) [inline services/create_project.py]: nice!

**Review by lukasab — CHANGES_REQUESTED** [review]
We can approve this after we solve the template activities issue.

**Comment by ribeirojose** [comment]
> > > I could not run the code because it needs the _template_ project. How are you thinking about creating this project(The one with all default activities)?
> > 
> > 
> > We should probably add this with the migrations, right? We should probably have a seed file to populate the db with this sort of thing. What do you think @lukasab ?
> 
> I think it does make a lot of sense adding this to the project/activities migration file. What do you mean by a seed file?

Something around these lines:

```python
from orator.seeds import Seeder
from functools import reduce

class ProjectTableSeeder(Seeder):

    def run(self):
        """
        Run the database seeds.
        """

        # Here you could just use random string generators
        # rather than hardcoded values
        self.db.table('project').insert({
           'id': 0,
            'name': 'template',
        }),
        default_proj_activity_ids = [ ... ]
        project_activity_content = reduce(
                lambda acc, id_: [*acc, {
                        'project_id': 0,
                        'activity_id': id_
                        }],
                        default_proj_activity_ids,
                        []) 
        self.db.table('project_activity').insert()
```

**Comment by lukasab** [comment]
> > > > I could not run the code because it needs the _template_ project. How are you thinking about creating this project(The one with all default activities)?
> > > 
> > > 
> > > We should probably add this with the migrations, right? We should probably have a seed file to populate the db with this sort of thing. What do you think @lukasab ?
> > 
> > 
> > I think it does make a lot of sense adding this to the project/activities migration file. What do you mean by a seed file?
> 
> Something around these lines:
> 
> ```python
> from orator.seeds import Seeder
> from functools import reduce
> 
> class ProjectTableSeeder(Seeder):
> 
>     def run(self):
>         """
>         Run the database seeds.
>         """
> 
>         self.db.table('project').insert({
>            'id': 0,
>             'name': 'template',
>         }),
>         default_activity_ids = [ ... ]
>         default_vals_project_activity = reduce(
>                 lambda acc, id_: [*acc, {
>                         'project_id': 0,
>                         'activity_id': id_
>                         }],
>                         default_activity_ids,
>                         []) 
>         self.db.table('project_activity').insert(default_vals_project_activity)
> ```

Have you implemented this?

**Comment by ribeirojose** [comment]
Closing due to no further activity.

**Pull request closed without merging**